### PR TITLE
[BUG FIX] Fix rigid body Jacobian getter for compound joints.

### LINKED ISSUE #1608: [Bug]: Wrong IK result for mobile manipulator
### Bug Description

Hi, I am using Genesis for mobile manipulation. I implemented a simple mobile manipulator (a base with a Franka arm). It works normally with joint space control. 

However, when I want to use the end effector Cartesian space PD control (e.g., I want the end effector to move forward for 0.01m and keep the base unmoved), the IK seems to return a wrong qpos. 

Specifically, in the following script, I have a mobile_franka robot, where the base has 3 dofs: moving left, right, and rotating, and the arm is a Franka Panda arm. I want to control the robot arm to move forward for `0.01m` every step and keep the base unmoved. I set the `arm_delta_action = torch.tensor([0.01, 0, 0, 0, 0, 0], device="cuda")`, and solve the IK to get the `target_qpos`, where I have set the `dofs_idx_local` to only count for the robot arm dofs (that is `[3,4,5,6,7,8,9]`). 

```
from copy import deepcopy
import sys

import torch
sys.path.append(".")
sys.path.append("..")
import numpy as np

import genesis as gs

from transform_utils import (
    euler_angles_to_matrix,
    matrix_to_quaternion,
    quaternion_multiply,
    quaternion_invert,
    Pose,
)

def get_pos(ee_link):
    return ee_link.get_pos()

def get_quat(ee_link):
    return ee_link.get_quat()

def get_pose(ee_link):
    return Pose.create_from_pq(ee_link.get_pos(), ee_link.get_quat())

def get_ee_link(entity, ee_link_name="hand"):
    return entity.get_link(ee_link_name)

gs.init(backend=gs.gpu)

scene = gs.Scene(
    viewer_options = gs.options.ViewerOptions(
        camera_pos    = (0, -3.5, 2.5),
        camera_lookat = (0.0, 0.0, 0.5),
        camera_fov    = 30,
        res           = (960, 640),
        max_FPS       = 60,
    ),
    sim_options = gs.options.SimOptions(
        dt = 0.01,
    ),
    show_viewer = True,
    show_FPS=True,
)

cam = scene.add_camera(
    res    = (960, 640),
    pos    = (0, -3.5, 2.5),
    lookat = (0.0, 0.0, 0.5),
    fov    = 30,
    GUI    = False
)

plane = scene.add_entity(
    gs.morphs.Plane(),
)
mobile_franka = scene.add_entity(
    gs.morphs.MJCF(
        file  = 'mobile/tidybot_panda.xml',
    ),
    material=gs.materials.Rigid(gravity_compensation=1.0),
)

scene.build()

scene.visualizer._context.add_light(
    {
        "type": "directional",  # point, directional, spot
        "dir": [0, 0.7071, -0.7071],  # Directional light pointing upwards
        "pos": [0, 0, 0],  # Position is not used for directional light
        "color": [1.0, 1.0, 1.0],
        "intensity": 6.0,
    }
)

jnt_names = [
    'joint_x',
    'joint_y',
    'joint_th',
    'joint1',
    'joint2',
    'joint3',
    'joint4',
    'joint5',
    'joint6',
    'joint7',
    'finger_joint1',
    'finger_joint2',
]
dofs_idx = [mobile_franka.get_joint(name).dof_idx_local for name in jnt_names]

mobile_franka.set_dofs_kp(
    kp             = np.array([600000, 600000, 30000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]),
    dofs_idx_local = dofs_idx,
)

mobile_franka.set_dofs_kv(
    kv             = np.array([90000, 90000, 5000, 100,  100,  100,  100,  100,  100,  100, 100, 100]),
    dofs_idx_local = dofs_idx,
)

mobile_franka.set_dofs_force_range(
    lower          = np.array([-10000, -10000, -10000, -100, -100, -100, -100, -100, -100, -100, -100, -100]),
    upper          = np.array([10000, 10000, 10000, 100,  100,  100,  100,  100,  100,  100,  100,  100]),    
    dofs_idx_local = dofs_idx,
)

cam.start_recording()

for i in range(10):
    # hard set to a good init pose
    mobile_franka.set_dofs_position(
        np.array([0, 0, 0, 0.0, -np.pi / 4, 0, -3 * np.pi / 4, 0, np.pi / 2, np.pi / 4, 0.04, 0.04]),
        dofs_idx,
    )
    scene.step()

ee_pose = get_pose(get_ee_link(mobile_franka))
init_eepose = deepcopy(ee_pose)

for i in range(200):
    prev_ee_pose = ee_pose

    arm_delta_action = torch.tensor([0.01, 0, 0., 0, 0, 0], device="cuda")   # xyzrpy

    delta_pos, delta_rot = arm_delta_action[:3], arm_delta_action[3:6]
    delta_quat = matrix_to_quaternion(euler_angles_to_matrix(delta_rot, "XYZ"))
    p = ee_pose.p + delta_pos
    q = quaternion_multiply(delta_quat, ee_pose.q)

    target_qpos = mobile_franka.inverse_kinematics(
        link=get_ee_link(mobile_franka),
        pos=p[0],
        # quat=q[0],
        quat=init_eepose.q[0],
        init_qpos=mobile_franka.get_dofs_position(dofs_idx),
        max_solver_iters=100,
        dofs_idx_local=dofs_idx[3:10],
        # rot_mask=[False, False, False]
        # pos_mask=[False, False, False],
    )
    mobile_franka.control_dofs_position(
        target_qpos,
        dofs_idx,
    )

    ee_pose = get_pose(get_ee_link(mobile_franka))

    print(f"Delta End effector pose: {ee_pose.p - prev_ee_pose.p}", f"{ee_pose.q - prev_ee_pose.q}")

    scene.step()
    cam.render()
    
cam.stop_recording(save_to_filename='video.mp4', fps=60)
```

The result is very strange. The robot arm starts to twist:

https://github.com/user-attachments/assets/86efed0f-f94e-422e-914f-f6c34d4533c9

I have one more experiment to show that it is indeed the problem of the IK:
If I set the action to all zero: `arm_delta_action = torch.tensor([0, 0, 0, 0, 0, 0], device="cuda")`, the `target_qpos` is correct, and the whole robot just keeps still.

### Steps to Reproduce

Scripts, utils, and model files are attached. Just run `test_genesis_mobile.py`.

[genesis_issue.zip](https://github.com/user-attachments/files/21891632/genesis_issue.zip)

### Expected Behavior

As stated above.

### Screenshots/Videos

_No response_

### Relevant log output

```shell

```

### Environment

- OS: Windows 11
- GPU RTX 4090

### Release version or Commit ID

commit ID: aa797a9, [BUG FIX] Fix Inverse Kinematics algorithm (cont'd).

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by vlordier**
## Root cause

`_func_get_jacobian` accumulated a per-link `dof_offset` and added it to the Jacobian column index:

```python
i_d_jac = i_d + dof_offset - self._dof_start
```

`joints_info.dof_start[I_j]` is the **global scene-level DOF index**, so subtracting `self._dof_start` already gives the correct entity-local column. Adding `dof_offset` on top of that shifts each successive joint's column by +1, +2, … causing later compound joints to overwrite earlier ones and producing wrong / discontinuous IK solutions.

**For the common case (one joint per link), `dof_offset` is always 0 — existing robots are unaffected.**  
The bug only triggers for **compound joints** (multiple revolute/prismatic joints on the same body), which is valid MJCF — a ball-socket joint is often modelled this way.

Example (2 revolute joints on `seg1`, DOFs 0 and 1):

| Joint | `i_d` | `dof_offset` | old `i_d_jac` | new `i_d_jac` |
|-------|-------|--------------|--------------|--------------|
| j_x   | 0     | 0            | 0 ✓          | 0 ✓           |
| j_y   | 1     | 1            | **2 ✗**      | 1 ✓           |
| j_z (child) | 2 | 0          | 2 (overwrites j_y!) | 2 ✓ |

## Fix

Remove `dof_offset` from the `i_d_jac` formula in all three joint-type branches (REVOLUTE, PRISMATIC, FREE):

```python
# Before
i_d_jac = i_d + dof_offset - self._dof_start

# After  
i_d_jac = i_d - self._dof_start
```

## Test

New fixture `xml/compound_joint.xml`: a 3-DOF arm where `seg1` carries two compound revolutes (`j_x`, `j_y`) and `seg2` carries one (`j_z`).  
`test_jacobian_compound_joints` verifies the analytical Jacobian at `q=[0,0,0]` against closed-form expected values. Max error = 0.0.

Related to #1608

### CODE REVIEW

**Comment by duburcqa** [comment]
@claude review

File: genesis/assets/xml/compound_joint.xml
Comment (duburcqa) [inline genesis/assets/xml/compound_joint.xml:1]: Remove this and use procedural generation.

File: tests/test_rigid_physics.py [hunk 1]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    """
+    Regression test for issue #1608: compound joints (multiple revolute joints on the same
+    body) produced wrong Jacobian columns because `dof_offset` was erroneously added to
+    `joints_info.dof_start`, which already encodes the global DOF position.
+
+    Before the fix, j_y (col 1) wrote to column 2 (1 + dof_offset=1 = 2) and j_z (col 2)
+    also wrote to column 2, making columns 1 and 2 identical. The test checks against the
+    known analytical Jacobian at q=[0,0,0] for a 3-DOF arm where seg1 has two compound
+    revolute joints (j_x, j_y) and seg2 has one (j_z).
+    """
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2945]: Remove docstring.

File: tests/test_rigid_physics.py [hunk 2]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    scene = gs.Scene(show_viewer=False)
+    robot = scene.add_entity(
+        gs.morphs.MJCF(file="xml/compound_joint.xml", requires_jac_and_IK=True),
+    )
+    scene.build()
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2950]: One line per options in unit tests and examples.

File: tests/test_rigid_physics.py [hunk 3]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    # Analytical Jacobian:
+    #   j_x (col 0): rot=[1,0,0], r=[0,0,0.4] → J_pos = r×ω = [0,-0.4,0], J_rot=[1,0,0]
+    #   j_y (col 1): rot=[0,1,0], r=[0,0,0.4] → J_pos = r×ω = [0.4,0,0],  J_rot=[0,1,0]
+    #   j_z (col 2): at seg2 origin; r=[0,0,0] → J_pos = [0,0,0],           J_rot=[0,0,1]
+    qpos = np.zeros(robot.n_dofs, dtype=gs.np_float)
+    robot.set_qpos(qpos, zero_velocity=True)
+    scene.step()
+
+    end_link = robot.get_link("seg2")
+    J = tensor_to_array(robot.get_jacobian(end_link))  # shape (6, 3)
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2962]: Remove `tensor_to_array`. It is useless here.

File: tests/test_rigid_physics.py [hunk 4]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    )
+    scene.build()
+
+    # Zero configuration: seg2 origin is at (0, 0, 0.4) in world frame.
+    # Analytical Jacobian:
+    #   j_x (col 0): rot=[1,0,0], r=[0,0,0.4] → J_pos = r×ω = [0,-0.4,0], J_rot=[1,0,0]
+    #   j_y (col 1): rot=[0,1,0], r=[0,0,0.4] → J_pos = r×ω = [0.4,0,0],  J_rot=[0,1,0]
+    #   j_z (col 2): at seg2 origin; r=[0,0,0] → J_pos = [0,0,0],           J_rot=[0,0,1]
+    qpos = np.zeros(robot.n_dofs, dtype=gs.np_float)
+    robot.set_qpos(qpos, zero_velocity=True)
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2958]: Remove this. It is already the default.

File: tests/test_rigid_physics.py [hunk 5]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+            [0.0, L, 0.0],   # pos x
+            [-L, 0.0, 0.0],  # pos y
+            [0.0, 0.0, 0.0], # pos z  (j_z at seg2 origin, no translation)
+            [1.0, 0.0, 0.0], # rot x
+            [0.0, 1.0, 0.0], # rot y
+            [0.0, 0.0, 1.0], # rot z
+        ],
+        dtype=gs.np_float,
+    )
+    assert_allclose(J, expected, tol=tol)
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2976]: Remove `np.array(` wrapping. it is necessary.

File: tests/test_rigid_physics.py [hunk 6]
```diff
@@ -2979,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    #   j_y (col 1): rot=[0,1,0], r=[0,0,0.4] → J_pos = r×ω = [0.4,0,0],  J_rot=[0,1,0]
+    #   j_z (col 2): at seg2 origin; r=[0,0,0] → J_pos = [0,0,0],           J_rot=[0,0,1]
+    qpos = np.zeros(robot.n_dofs, dtype=gs.np_float)
+    robot.set_qpos(qpos, zero_velocity=True)
+    scene.step()
+
+    end_link = robot.get_link("seg2")
+    J = tensor_to_array(robot.get_jacobian(end_link))  # shape (6, 3)
+
+    L = 0.4  # seg1 length
```
Comment (duburcqa) [inline tests/test_rigid_physics.py:2964]: Remove this constant, it is useless.

**Review by duburcqa — APPROVED** [review]

File: genesis/engine/entities/rigid_entity/rigid_entity.py [hunk 7]
```diff
@@ -2482,7 +2481,9 @@ def _func_get_jacobian(
                 elif joints_info.type[I_j] == gs.JOINT_TYPE.REVOLUTE:
                     i_d = joints_info.dof_start[I_j]
                     I_d = [i_d, i_b] if qd.static(self.solver._options.batch_dofs_info) else i_d
-                    i_d_jac = i_d + dof_offset - self._dof_start
+                    # joints_info.dof_start is the global scene-level DOF index; subtract entity
+                    # offset to get the entity-local column index into _jacobian.
+                    i_d_jac = i_d - self._dof_start
                     rotation = gu.qd_transform_by_quat(dofs_info.motion_ang[I_d], links_state.quat[i_l, i_b])
                     translation = rotation.cross(tgt_link_pos - links_state.pos[i_l, i_b])
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/entities/rigid_entity/rigid_entity.py]: Fixed: the REVOLUTE and PRISMATIC branches now read `joints_state.xaxis` and `joints_state.xanchor` directly, which carry the correct per-joint pre-rotation pose for compound joints.

File: tests/test_rigid_physics.py [hunk 8]
```diff
@@ -2931,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    # Analytical Jacobian:
+    #   j_x (col 0): rot=[1,0,0], r=[0,0,0.4] → J_pos = r×ω = [0,-0.4,0], J_rot=[1,0,0]
+    #   j_y (col 1): rot=[0,1,0], r=[0,0,0.4] → J_pos = r×ω = [0.4,0,0],  J_rot=[0,1,0]
+    #   j_z (col 2): at seg2 origin; r=[0,0,0] → J_pos = [0,0,0],           J_rot=[0,0,1]
+    qpos = np.zeros(robot.n_dofs, dtype=gs.np_float)
+    robot.set_qpos(qpos, zero_velocity=True)
+    scene.step()
+
+    end_link = robot.get_link("seg2")
+    J = tensor_to_array(robot.get_jacobian(end_link))  # shape (6, 3)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Added a `q=[0.3, -0.5, 0.7]` case validated against `mujoco.mj_jacBody` as ground truth, which exercises the compound-joint axis path.

File: tests/test_rigid_physics.py [hunk 9]
```diff
@@ -2931,6 +2931,51 @@ def test_jacobian(gs_sim, tol):
...
+    robot = scene.add_entity(
+        gs.morphs.MJCF(file="xml/compound_joint.xml", requires_jac_and_IK=True),
+    )
+    scene.build()
+
+    # Zero configuration: seg2 origin is at (0, 0, 0.4) in world frame.
+    # Analytical Jacobian:
+    #   j_x (col 0): rot=[1,0,0], r=[0,0,0.4] → J_pos = r×ω = [0,-0.4,0], J_rot=[1,0,0]
+    #   j_y (col 1): rot=[0,1,0], r=[0,0,0.4] → J_pos = r×ω = [0.4,0,0],  J_rot=[0,1,0]
+    #   j_z (col 2): at seg2 origin; r=[0,0,0] → J_pos = [0,0,0],           J_rot=[0,0,1]
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/test_rigid_physics.py]: The analytical-Jacobian comment block was removed entirely in favor of a MuJoCo ground-truth comparison.

File: genesis/assets/xml/compound_joint.xml
Comment (duburcqa) [reply genesis/assets/xml/compound_joint.xml:1]: Done: the asset is deleted and replaced by a `compound_joint` session fixture built with `ET.Element`.

File: tests/test_rigid_physics.py (hunk 1 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Done.

File: tests/test_rigid_physics.py (hunk 2 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py:2960]: Done.

File: tests/test_rigid_physics.py (hunk 3 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Done.

File: tests/test_rigid_physics.py (hunk 4 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Done.

File: tests/test_rigid_physics.py (hunk 5 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Done.

File: tests/test_rigid_physics.py (hunk 6 above)
Comment (duburcqa) [reply tests/test_rigid_physics.py]: Done.
